### ISSUE TITLE
Do not recompile the sketch if there are no changes and libraries have not changed

### ISSUE THREAD

**Issue by cmaglie** [arduino/arduino-cli#2204]
### Describe the request

If a sketch is recompiled twice in a row, the Arduino-preprocessed sketch `Sketch.ino.cpp` is recompiled and re-linked even if not necessary.

### Describe the current behavior

If you do the same compile twice in a row:

```
~/Arduino/Blink$ LANG=en arduino-cli compile -b arduino:avr:uno --clean
Sketch uses 1636 bytes (5%) of program storage space. Maximum is 32256 bytes.
Global variables use 122 bytes (5%) of dynamic memory, leaving 1926 bytes for local variables. Maximum is 2048 bytes.

Used library Version Path                                                                       
Wire         1.0     /home/megabug/.arduino15/packages/arduino/hardware/avr/1.8.6/libraries/Wire

Used platform Version Path                                                        
arduino:avr   1.8.6   /home/megabug/.arduino15/packages/arduino/hardware/avr/1.8.6

~/Arduino/Blink$ LANG=en arduino-cli compile -b arduino:avr:uno -v
FQBN: arduino:avr:uno
Using board 'uno' from platform in folder: /home/megabug/.arduino15/packages/arduino/hardware/avr/1.8.6
Using core 'arduino' from platform in folder: /home/megabug/.arduino15/packages/arduino/hardware/avr/1.8.6

Detecting libraries used...
[...]
Generating function prototypes...
[...]
Compiling sketch...
/home/megabug/.arduino15/packages/arduino/tools/avr-gcc/7.3.0-atmel3.6.1-arduino7/bin/avr-g++ -c -g -Os -w -std=gnu++11 -fpermissive -fno-exceptions -ffunction-sections -fdata-sections -fno-threadsafe-statics -Wno-error=narrowing -MMD -flto -mmcu=atmega328p -DF_CPU=16000000L -DARDUINO=10607 -DARDUINO_AVR_UNO -DARDUINO_ARCH_AVR -I/home/megabug/.arduino15/packages/arduino/hardware/avr/1.8.6/cores/arduino -I/home/megabug/.arduino15/packages/arduino/hardware/avr/1.8.6/variants/standard -I/home/megabug/.arduino15/packages/arduino/hardware/avr/1.8.6/libraries/Wire/src /tmp/arduino/sketches/002050EAA7EFB9A4FC451CDFBC0FA2D3/sketch/Blink.ino.cpp -o /tmp/arduino/sketches/002050EAA7EFB9A4FC451CDFBC0FA2D3/sketch/Blink.ino.cpp.o
Compiling libraries...
Compiling library "Wire"
Using previously compiled file: /tmp/arduino/sketches/002050EAA7EFB9A4FC451CDFBC0FA2D3/libraries/Wire/utility/twi.c.o
Using previously compiled file: /tmp/arduino/sketches/002050EAA7EFB9A4FC451CDFBC0FA2D3/libraries/Wire/Wire.cpp.o
Compiling core...
Using precompiled core: /tmp/arduino/cores/arduino_avr_uno_12e4cfbdc0590d50ab9cd20e50a4c3c5/core.a
Linking everything together...
/home/megabug/.arduino15/packages/arduino/tools/avr-gcc/7.3.0-atmel3.6.1-arduino7/bin/avr-gcc -w -Os -g -flto -fuse-linker-plugin -Wl,--gc-sections -mmcu=atmega328p -o /tmp/arduino/sketches/002050EAA7EFB9A4FC451CDFBC0FA2D3/Blink.ino.elf /tmp/arduino/sketches/002050EAA7EFB9A4FC451CDFBC0FA2D3/sketch/Blink.ino.cpp.o /tmp/arduino/sketches/002050EAA7EFB9A4FC451CDFBC0FA2D3/libraries/Wire/Wire.cpp.o /tmp/arduino/sketches/002050EAA7EFB9A4FC451CDFBC0FA2D3/libraries/Wire/utility/twi.c.o /tmp/arduino/sketches/002050EAA7EFB9A4FC451CDFBC0FA2D3/../../cores/arduino_avr_uno_12e4cfbdc0590d50ab9cd20e50a4c3c5/core.a -L/tmp/arduino/sketches/002050EAA7EFB9A4FC451CDFBC0FA2D3 -lm
/home/megabug/.arduino15/packages/arduino/tools/avr-gcc/7.3.0-atmel3.6.1-arduino7/bin/avr-objcopy -O ihex -j .eeprom --set-section-flags=.eeprom=alloc,load --no-change-warnings --change-section-lma .eeprom=0 /tmp/arduino/sketches/002050EAA7EFB9A4FC451CDFBC0FA2D3/Blink.ino.elf /tmp/arduino/sketches/002050EAA7EFB9A4FC451CDFBC0FA2D3/Blink.ino.eep
/home/megabug/.arduino15/packages/arduino/tools/avr-gcc/7.3.0-atmel3.6.1-arduino7/bin/avr-objcopy -O ihex -R .eeprom /tmp/arduino/sketches/002050EAA7EFB9A4FC451CDFBC0FA2D3/Blink.ino.elf /tmp/arduino/sketches/002050EAA7EFB9A4FC451CDFBC0FA2D3/Blink.ino.hex
Multiple libraries were found for "Wire.h"
  Used: /home/megabug/.arduino15/packages/arduino/hardware/avr/1.8.6/libraries/Wire
  Not used: /home/megabug/Workspace/Arduino_Threads
Using library Wire at version 1.0 in folder: /home/megabug/.arduino15/packages/arduino/hardware/avr/1.8.6/libraries/Wire 
/home/megabug/.arduino15/packages/arduino/tools/avr-gcc/7.3.0-atmel3.6.1-arduino7/bin/avr-size -A /tmp/arduino/sketches/002050EAA7EFB9A4FC451CDFBC0FA2D3/Blink.ino.elf
Sketch uses 1636 bytes (5%) of program storage space. Maximum is 32256 bytes.
Global variables use 122 bytes (5%) of dynamic memory, leaving 1926 bytes for local variables. Maximum is 2048 bytes.

Used library Version Path                                                                       
Wire         1.0     /home/megabug/.arduino15/packages/arduino/hardware/avr/1.8.6/libraries/Wire

Used platform Version Path                                                        
arduino:avr   1.8.6   /home/megabug/.arduino15/packages/arduino/hardware/avr/1.8.6
```

🐛 The sketch file `Blink.ino.cpp` is recompiled even if there are no changes.

The sketch and linking phase could be completely skipped in this particular case.

### Arduino CLI version

nightly

### Operating system

N/A

### Operating system version

N/A

#### Additional requests

- https://github.com/arduino/arduino-cli/issues/1996
- https://github.com/arduino/arduino-cli/issues/1996#issuecomment-1440479931
- https://github.com/arduino/arduino-cli/issues/1996#issuecomment-1585813298
- https://github.com/arduino/arduino-cli/issues/2015
- https://github.com/arduino/arduino-ide/issues/270
- https://github.com/arduino/arduino-ide/issues/1789
- https://github.com/arduino/arduino-ide/issues/2103
- https://github.com/arduino/arduino-ide/issues/2140
- https://github.com/arduino/arduino-ide/issues/2732
- https://github.com/arduino/Arduino/issues/2335
- https://github.com/arduino/Arduino/issues/6625
- https://github.com/arduino/arduino/issues/8149
- https://forum.arduino.cc/t/recompiling-on-upload-without-code-change/897394
- https://forum.arduino.cc/t/arduino-ide-compiles-every-time-and-is-slow/1044606
- https://forum.arduino.cc/t/ide-2-0-2-why-recompile-every-time/1064264
- https://forum.arduino.cc/t/upload-without-compile/1175639
- https://forum.arduino.cc/t/still-no-upload-without-recompile/1300242

### Issue checklist

- [X] I searched for previous requests in [the issue tracker](https://github.com/arduino/arduino-cli/issues?q=)
- [X] I verified the feature was still missing when using the [nightly build](https://arduino.github.io/arduino-cli/dev/installation/#nightly-builds)
- [X] My request contains all necessary details

**Comment by cmaglie** [comment]
Should be fixed by #3112

**Comment by tempelmann** [comment]
I just tried out the latest build (Version: 2.3.9-nightly-20260426, Date: 2026-04-26T04:16:23.286Z (1 day ago), CLI Version: 1.4.1)

But I can't see any improvement in regards to build time for a .ino project that uses libs and extra local .h/.cpp files. Isn't this meant to deal with the long-standing issue that a second Verify won't recompile code again?

I'm referring to the discussion in #1996 where it was explained:

> Actually the compile process does compile different "parts" of a sketch:
> • the core
> • the libraries
> • the user code (the sketch)
> The first two already use a caching mechanism. The sketch currently does not have this behavior. Unfortunately, it's not that easy because the library detection can't be cached.

I had a look at the generated files, and see that there are already several "helper" files that can be used for caching, so I wonder what the difficulty is to resolve this.

Here's my take on in:

Can't you simply look at (and cache) the file modification times, just like "make" does it?

I mean that you could skip the entire library import detection process if you see that the _source_ files have not been changed since the last time the build files were processed or written.

In other words, if the sources have not changed since the last compile, why isn't that enough to skip the entire processing?

**Comment by per1234** [comment]
>CLI Version: 1.4.1

@tempelmann the change was made after the Arduino CLI 1.4.1 release, so it is expected that you would not see its effect when using Arduino CLI 1.4.1. The Arduino CLI dependency of Arduino IDE is pinned, and is only bumped periodically by the Arduino IDE developers, so you won't be using the latest development version of Arduino CLI even when using the nightly build of Arduino IDE.

You should use the nightly build of Arduino CLI directly if you wish to evaluate the effects of https://github.com/arduino/arduino-cli/pull/3112:

https://arduino.github.io/arduino-cli/dev/installation/#nightly-builds

If you have any questions or problems with using the nightly build please post over on Arduino Forum:

https://forum.arduino.cc/c/development-tools/arduino-command-line-tools/89

We'll be happy to provide assistance over there. I closely monitor the forum, so will be sure to see any post you might make.

**Comment by tempelmann** [comment]
Thank you. See my related question in the forum, though: https://forum.arduino.cc/t/how-do-i-use-the-latest-arduino-cli-with-the-current-ide/1441731
